### PR TITLE
Eliminate lock contention in `ConcurrentHashJoin`

### PR DESCRIPTION
**Pull request by lgbo-ustc**
### Changelog category (leave one):

- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

By the following test case, We found there exists intense lock contention in `ConcurrentHashJoin`.
```sql
CREATE TABLE default.test_data_left
(
    `id` UInt64,
    `x` UInt64,
    `y` UInt64,
    `z` UInt64,
    `w` String
)
ENGINE = Memory
insert into default.test_data_left select number as id, number % 1111 as x, number * 3 as y, number - 100 as z, cast(number as String) as w from numbers(100000000);

CREATE TABLE default.test_data_right
(
    `id` UInt64,
    `x` UInt64,
    `y` UInt64,
    `z` UInt64,
    `w` String
)
ENGINE = Memory
insert into default.test_data_right select number as id, number % 1111 as x, number * 3 as y, number - 100 as z, cast(number as String) as w from numbers(100000000);

select t1.id, t1.x, t1.y, t2.id, t2.x, t2.y from test_data_left as t1 any inner join test_data_right as t2 on t1.id = t2.id FORMAT Null settings max_threads=8, join_algorithm='parallel_hash'

```
the `perf top` as following
```
Samples: 860K of event 'cycles', 4000 Hz, Event count (approx.): 197328586134 lost: 0/0 drop: 0/132
Overhead  Shared Object       Symbol
  36.56%  clickhouse          [.] DB::JoinStuff::JoinUsedFlags::reinit<(DB::JoinKind)0, (DB::JoinStrictness)2>
  25.11%  clickhouse          [.] DB::(anonymous namespace)::joinRightColumns<(DB::JoinKind)0, (DB::JoinStrictness)2, DB
   7.22%  libpthread-2.31.so  [.] pthread_mutex_trylock
   5.71%  clickhouse          [.] DB::(anonymous namespace)::insertFromBlockImplTypeCase<(DB::JoinStrictness)2, DB::Colu
   1.96%  clickhouse          [.] HashTable<unsigned long, HashMapCell<unsigned long, DB::RowRef, HashCRC32<unsigned lon
   1.77%  clickhouse          [.] memcpy
   1.76%  clickhouse          [.] DB::ColumnVector<unsigned long>::insertFrom
   1.69%  clickhouse          [.] DB::IColumn::scatterImpl<DB::ColumnVector<unsigned long> >
   1.55%  clickhouse          [.] std::__1::__hash_table<std::__1::__hash_value_type<DB::Block const*, std::__1::vector<
   1.36%  clickhouse          [.] DB::ConcurrentHashJoin::addJoinedBlock
   1.17%  [kernel]            [k] smp_call_function_many
   0.85%  clickhouse          [.] DB::(anonymous namespace)::AddedColumns::appendFromBlock<false>
   0.76%  [kernel]            [k] native_irq_return_iret
   0.62%  clickhouse          [.] DB::Block::allocatedBytes
   0.61%  [kernel]            [k] __default_send_IPI_dest_field
   0.59%  [kernel]            [k] llist_add_batch
   0.55%  [kernel]            [k] clear_page_erms
   0.49%  [kernel]            [k] flush_smp_call_function_queue
```
```
0 rows in set. Elapsed: 13.900 sec. Processed 200.00 million rows, 4.80 GB (14.39 million rows/s., 345.33 MB/s.)
```

In this pr, we make partition shuffle for both `left` and `right` before joining, this can eliminate the lock but make the `scatter` cost remain the same as before.

Here we have the optimization result.
```
Samples: 617K of event 'cycles', 4000 Hz, Event count (approx.): 115228078214 lost: 0/0 drop: 0/17
Overhead  Shared Object         Symbol
  39.72%  clickhouse            [.] DB::(anonymous namespace)::joinRightColumns<(DB::JoinKind)0, (DB::JoinStrictness)2,
  10.35%  clickhouse            [.] DB::JoinStuff::JoinUsedFlags::reinit<(DB::JoinKind)0, (DB::JoinStrictness)2>
   9.96%  clickhouse            [.] DB::(anonymous namespace)::insertFromBlockImplTypeCase<(DB::JoinStrictness)2, DB::Co
   3.73%  clickhouse            [.] DB::IColumn::scatterImpl<DB::ColumnVector<unsigned long> >
   3.41%  clickhouse            [.] memcpy
   2.97%  clickhouse            [.] DB::ColumnVector<unsigned long>::insertFrom
   2.88%  clickhouse            [.] HashTable<unsigned long, HashMapCell<unsigned long, DB::RowRef, HashCRC32<unsigned l
   2.56%  clickhouse            [.] std::__1::__hash_table<std::__1::__hash_value_type<DB::Block const*, std::__1::vecto
   2.13%  [kernel]              [k] smp_call_function_many
   1.61%  clickhouse            [.] DB::(anonymous namespace)::AddedColumns::appendFromBlock<false>
   1.35%  [kernel]              [k] native_irq_return_iret
   0.97%  [kernel]              [k] llist_add_batch
   0.97%  [kernel]              [k] clear_page_erms
   0.90%  [kernel]              [k] __default_send_IPI_dest_field
   0.83%  [kernel]              [k] _raw_spin_lock
   0.66%  [kernel]              [k] llist_reverse_order
   0.64%  [kernel]              [k] flush_smp_call_function_queue
   0.56%  clickhouse            [.] DB::ColumnVector<unsigned long>::updateWeakHash32
   0.54%  [kernel]              [k] sync_regs
   0.54%  [kernel]              [k] mem_cgroup_try_charge
   0.50%  [kernel]              [k] down_read_trylock
   0.50%  [kernel]              [k] native_flush_tlb
   0.47%  [kernel]              [k] __handle_mm_fault

```

```
0 rows in set. Elapsed: 8.296 sec. Processed 200.00 million rows, 4.80 GB (24.11 million rows/s., 578.60 MB/s.)
```

It's about +40% speed up in the above case, but it will be lower in actual use. 

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by lgbo-ustc** [comment]
There are still some cases we need to fix.

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit 58a23f5297663821e9d43ca1e5c6d466eaa03da2 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/47638/58a23f5297663821e9d43ca1e5c6d466eaa03da2/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Interpreters/ConcurrentHashJoin.h [hunk 1]
```diff
@@ -1,40 +1,27 @@
...
-#include <base/types.h>
-#include <Common/Stopwatch.h>
-
+#include <Interpreters/Context_fwd.h>
+#include <Interpreters/HashJoin.h>
+#include "Context.h"
 namespace DB
 {
 
-/**
```
Comment (devcrafter) [inline src/Interpreters/ConcurrentHashJoin.h:18]: Why have you deleted the algorithm description? If the behavior is changed within this PR, please update the comment accordingly

**Comment by lgbo-ustc** [comment]
> @lgbo-ustc It looks like the PR is ready for review, right?

Not ready, the pipeline still get suck on some corner cases

**Pull request closed without merging**